cel/library.go

```diff
@@ -439,135 +438,1241 @@ func (l library) validateEmail(addr string) bool {
+				return true
+			}
+			return false
+		}
+		if i.h16() {
+			continue
+		}
+		if i.take(':') { //nolint:nestif
+			if i.take(':') {
+				if i.doubleColonSeen {
+					return false
+				}
+				i.doubleColonSeen = true
+				i.doubleColonAt = len(i.pieces)
+				if i.take(':') {
+					return false
+				}
+			}
+			continue
+		}
+		if i.str[i.index] == '%' && !i.zoneID() {
+			return false
+		}
+		break
+	}
+	return i.doubleColonSeen || len(i.pieces) == 8
+}
+
+// There is no definition for the character set allowed in the zone
+// identifier. RFC 4007 permits basically any non-null string.
+//
+// RFC 6874 defines it as ZoneID = 1*( unreserved / pct-encoded ).
+func (i *ipv6) zoneID() bool {
+	start := i.index
+	if i.take('%') {
+		if i.strLen-i.index > 0 {
+			// permit any non-null string
+			i.index = i.strLen
+			i.zoneIDFound = true
+			return true
+		}
+	}
+	i.index = start
+	i.zoneIDFound = false
+	return false
+}
+
+// Parses str from the current index to determine if it contains a valid
+// dotted sequence defined as:
+//
+// 1*3DIGIT "." 1*3DIGIT "." 1*3DIGIT "." 1*3DIGIT
+// Stores match in `dottedRaw`.
+func (i *ipv6) dotted() bool {
+	start := i.index
+	i.dottedRaw = ""
+	for {
+		if i.index < i.strLen && (i.digit() || i.take('.')) {
+			continue
+		}
+		break
+	}
+	if i.index-start >= 7 {
+		i.dottedRaw = i.str[start:i.index]
+		return true
+	}
+	i.index = start
+	return false
+}
+
+// h16 = 1*4HEXDIG
+// Stores 16-bit value in `pieces`.
+func (i *ipv6) h16() bool {
+	start := i.index
+	for {
+		if i.index >= i.strLen || !i.hexdig() {
+			break
+		}
+	}
+	str := i.str[start:i.index]
+	if len(str) == 0 {
+		// too short
+		return false
+	}
+	if len(str) > 4 {
+		// too long
+		return false
+	}
+
+	value, err := strconv.ParseUint(str, 16, 16)
+	if err != nil {
+		return false
+	}
+	i.pieces = append(i.pieces, uint16(value))
+	return true
+}
+
+// HEXDIG =  DIGIT / "A" / "B" / "C" / "D" / "E" / "F".
+func (i *ipv6) hexdig() bool {
+	c := i.str[i.index]
+	if ('0' <= c && c <= '9') ||
+		('a' <= c && c <= 'f') ||
+		('A' <= c && c <= 'F') {
+		i.index++
+		return true
+	}
+	return false
+}
+
+// DIGIT = %x30-39  ; 0-9.
+func (i *ipv6) digit() bool {
+	c := i.str[i.index]
+	if '0' <= c && c <= '9' {
+		i.index++
+		return true
+	}
+	return false
+}
+
+// If char is at the current index, return true and increment the index.
+// If char is not at the current index or the end of str has been reached,
+// return false.
+func (i *ipv6) take(char byte) bool {
+	if i.index >= i.strLen {
+		return false
+	}
+	if i.str[i.index] == char {
+		i.index++
+		return true
+	}
+	return false
+}
+
+// newIpv6 creates a new ipv6 based on str.
+func newIpv6(str string) *ipv6 {
+	return &ipv6{
+		str:           str,
+		strLen:        len(str),
+		pieces:        make([]uint16, 0),
+		doubleColonAt: -1,
+		dottedAddr:    nil,
+	}
+}
+
+/**
+ * Returns true if the string is an IPv4 or IPv6 address, optionally limited to
+ * a specific version.
```

Comment:
I'm not sure I've ever seen block comments in Go code, and would have expected line comments here. Deferring to @rodaine, @jchadwick-buf.

Reply:
Line comments only please.

Reply:
The only problem is that the linter complains about invalid formatting if you try to align things according to the RFC grammar with line comments (see `hierPart` for an example). 

What is the Go standard practice for getting around that @rodaine? Other than adding an ignore I guess.

Reply:
Check out the doc comment (er...) [docs on code blocks](https://tip.golang.org/doc/comment#code).

Reply:
i'm gonna guess it's the tabbing on 1142. Does running `golangci-lint run --fix` change it?

Reply:
Yeah it mangles it even further. The best I can do it seems to have a line break in between. Go seems to really want that to happen so 🤷

Reply:
Fixing the formatting: https://github.com/bufbuild/protovalidate-go/pull/208